examples/matmul.py

```diff
@@ -44,32 +44,35 @@ def main():
             c_part.append(cp.empty((c_dim, n), np.float32, order = 'F'))
 
     previous = None
-    for repetition in range(repetitions):
-        # Now compute a @ b.T and write the output to c
-        deps = [previous] if previous is not None else []
-        @spawn(placement = cpu, dependencies = deps)
-        async def run_matmul():
-            print("start")
-            start = time.perf_counter()
-            matmul = TaskSpace("matmul")
-            for i in range(blocks):
-                for j in range(blocks):
-                    a_block = a_part[i]
-                    b_block = b_part[j]
-                    c_block = c_part[i][:, j * block_size : (j + 1) * block_size]
-                    @spawn(matmul[i, j], placement = c_block)
-                    def matmul_task():
-                        old_device = cp.cuda.Device()
-                        #b_block_local = clone_here(b_block)
-                        b_block_local = cp.asarray(b_block, order = 'F')
-                        # cupy doesn't support the out argument for matmul yet so we have to copy.
-                        # cp.matmul(a_block, b_block_local.T, out = c_block)
-                        #print(i, j, old_device, cp.cuda.Device(), c_block.device, a_block.device, b_block_local.device, b_block.device)
-                        c_block[:] = a_block @ b_block_local.T
-            await matmul
-            stop = time.perf_counter()
-            print(stop - start)
-        previous = run_matmul
+    with reserve_persistent_memory([a_part, b_part, c_part]):
+        for repetition in range(repetitions):
+            # Now compute a @ b.T and write the output to c
+            deps = [previous] if previous is not None else []
+            @spawn(placement = cpu, dependencies = deps)
+            async def run_matmul():
+                start = time.perf_counter()
+                matmul = TaskSpace("matmul")
+                for i in range(blocks):
+                    for j in range(blocks):
+                        a_block = a_part[i]
+                        b_block = b_part[j]
+                        c_block = c_part[i][:, j * block_size : (j + 1) * block_size]
+                        memsize = c_block.nbytes
+                        if i != j:
+                            memsize += b_block.nbytes
+                        @spawn(matmul[i, j], placement = c_block, memory = memsize)
```

Comment:
We should make `storage_size` remove duplicate or overlapping arrays from it's argument list, so this can just be `memory = storage_size(c_block, b_block)`.

Reply:
I agree with that suggestion. In this case it wouldn't help though since the issue is that if b_block is already on the current device space was reserved for it elsewhere. Idk what a good fix for that would be. We could track reserved arrays somewhere, but I'm hesitant to add a global buffer pool like that.

Reply:
Hmm. OK. I think this is just another reason that we really need a PTask-like buffer tracking system, so that we can handle all this as part of the cache protocol.